examples/autoignition-mpi.py

```diff
@@ -233,23 +236,37 @@ def my_checkpoint(step, t, dt, state):
         cv = split_conserved(dim, state)
         reaction_rates = eos.get_production_rates(cv)
         viz_fields = [("reaction_rates", reaction_rates)]
-        return sim_checkpoint(discr, visualizer, eos, q=state,
-                              vizname=casename, step=step,
-                              t=t, dt=dt, nstatus=nstatus, nviz=nviz,
-                              constant_cfl=constant_cfl, comm=comm,
+        try:
+            # Check the health of the simulation
+            cfd_healthcheck(discr, eos, state,
+                            step=step, t=t, freq=ncheck)
+            # Perform checkpointing
+            sim_checkpoint(discr, eos, state,
+                           step=step, t=t, dt=dt, freq=nstatus,
+                           constant_cfl=constant_cfl)
+            # Visualize
+            sim_visualization(discr, eos, state,
+                              visualizer, vizname=casename,
+                              step=step, t=t, freq=nviz,
+                              viz_fields=viz_fields)
+        except StepperCrashError as err:
```

Comment:
@majosm @MTCam here's the updated proposal

Reply:
Cool, thanks @thomasgibson.  I like how you've split these functions out.  This work interacts strongly with that of #257.

 - I dunno about StepperCrashError for these utilities. I do see @majosm's comment about this earlier, but these aren't really stepper errors. I think it is a little confusing to toss stepper errors from these functions that are (on the surface) unrelated to the actual stepping and stepper.
 - I'm not a giant fan of having these utilities check whether they will do something on the inside (via step/freq).  I think it is a fairly good policy to make the routine very dumb; if it is called, it does the thing.  Let the caller worry about if it is time to invoke the function. The upshot/question is: Any objections to putting the `check_step` call on the *outside*? 
 - I would be in favor of renaming these utilities.  I like `cfd_healthcheck` but would maybe change that to `fluid_healthcheck`. The other two are fluid-specific too. As long as we're changing them, we may as well change `sim_visualization` to `fluid_visualization`.  
 - `sim_checkpoint` is basically eliminated by this change; we can go back later and reshape it to be general in #257, or banish it forever.  If we banish, we need to add soln watch via `logpyle` to the examples that lack it.

Reply:
We do need to write some tests that expose some of the stuff that we are concerned about for error handling.  I don't have any specific good ideas about how to test these atm, but some of the stuff we've talked about:

- Asynchronous errors during stepping
- Viz file collisions/overwrites
- Graceful exit after any exceptions
- Handling exact soln mismatch collectively

Some of this may be firmly in #257 territory.

Reply:
To give some background here @thomasgibson: the main reason `sim_checkpoint` exists is to reduce the amount of duplicated code in the drivers. Ideally we want the code inside `my_checkpoint` to be pretty simple (unless the user wants to do something extra that's specific to their simulation). So, with that motivation in mind, I wonder if there's a way we could move the exception handling part inside `sim_checkpoint`? Maybe something along the lines of:

```python
# driver.py

def healthcheck(step, t, state):
    return cfd_healthcheck(...)

def write_vis(step, t, state):
    return sim_visualization(...)

def checkpoint(step, t, dt, state):
    return sim_checkpoint(..., healthcheck=healthcheck, write_vis=write_vis)
```

```python
# simutil.py

def sim_checkpoint(..., healthcheck=None, write_vis=None):
    try:
        if check_step(...):
            # ... write status ...

        if check_step(...) and healthcheck is not None:
            healthcheck(step, t, state)

        if check_step(...) and write_vis is not None:
            write_vis(step, t, state)

    except StepperCrashError as e:
        # ... log error ...

        if write_vis is not None:
            write_vis(step, t, state)

        raise
```

> * I dunno about StepperCrashError for these utilities. I do see @majosm's comment about this earlier, but these aren't really stepper errors. I think it is a little confusing to toss stepper errors from these functions that are (on the surface) unrelated to the actual stepping and stepper.

IMO the name made sense when the `try`/`except` was wrapped around `advance_state`, but now that it's inside the checkpoint I think it could be named something else. Maybe the base class can just be something low level that indicates that it's raised globally? `SynchronizedError`? 🤷‍♂️ 

> * I'm not a giant fan of having these utilities check whether they will do something on the inside (via step/freq).  I think it is a fairly good policy to make the routine very dumb; if it is called, it does the thing.  Let the caller worry about if it is time to invoke the function. The upshot/question is: Any objections to putting the `check_step` call on the _outside_?

I think this depends on whether the code in question is designed to maximize flexibility or simplicity. `sim_checkpoint` is more intended to do the latter, so I think it makes sense to wrap the `check_step` calls inside it.

> * I would be in favor of renaming these utilities.  I like `cfd_healthcheck` but would maybe change that to `fluid_healthcheck`. The other two are fluid-specific too. As long as we're changing them, we may as well change `sim_visualization` to `fluid_visualization`.

I think I would vote for just calling it `sim_healthcheck` for the time being and fixing it in #257.

Reply:
> I think this depends on whether the code in question is designed to maximize flexibility or simplicity. `sim_checkpoint` is more intended to do the latter, so I think it makes sense to wrap the `check_step` calls inside it.
> 

imo, removing the check both simplifies and increases flexibility (the way it was before). I agree that if the `sim_checkpoint` ends up wrapping all the checks, that `sim_checkpoint` needs to perform the step checks, but I'd still argue that the call to `check_step` should be at the call site to `sim_healthcheck` and not down inside of it. But... this is also a very mild and minor complaint.  It has been down inside the routine all this time.

> > * I would be in favor of renaming these utilities.  I like `cfd_healthcheck` but would maybe change that to `fluid_healthcheck`. The other two are fluid-specific too. As long as we're changing them, we may as well change `sim_visualization` to `fluid_visualization`.
> 
> I think I would vote for just calling it `sim_healthcheck` for the time being and fixing it in #257.

👍

Reply:
@majosm @MTCam I've pushed a change that renames `cfd_healthcheck`, and I've renamed the `StepperCrashError` according to @majosm's comment: https://github.com/illinois-ceesd/mirgecom/pull/370/commits/2523fbdfa581c0c86f0e4d21691aa345a64d185b.

Is there anything else you'd like for me to apply before leaving the rest to https://github.com/illinois-ceesd/mirgecom/pull/257?

Reply:
Is there a good reason that `sim_checkpoint` does not make a crash viz dump at exception time for serial runs?

I'm not going to pursue this complaint very far - but I think that we are going backwards a bit here.  This works essentially strips the user of a good portion of the control over what happens when the health check fails.   Health checks are meant to be a user-defined facility.  Imo, *MIRGE-Com* should provide simple utilities to perform health checks, but it should be up to the user to handle what happens if any of those checks are failed.   For me, it is a step backwards to roll a new monolithic "checkpoint driver" that tries to do and automate too much.

Reply:
> Is there a good reason that `sim_checkpoint` does not make a crash viz dump at exception time for serial runs?

?

> I'm not going to pursue this complaint very far - but I think that we are going backwards a bit here. This works essentially strips the user of a good portion of the control over what happens when the health check fails. Health checks are meant to be a user-defined facility. Imo, _MIRGE-Com_ should provide simple utilities to perform health checks, but it should be up to the user to handle what happens if any of those checks are failed. For me, it is a step backwards to roll a new monolithic "checkpoint driver" that tries to do and automate too much.

Sounds like we don't have a consensus about what we're actually trying to do here... i.e., are we trying to maximize user control, or minimize code? They conflict with each other to an extent. I was under the impression that we were aiming more for the latter.

I think it would make sense to have the user pass in a health check function to `sim_checkpoint`. I'm less sure I see the value in having them handle errors manually. Can you give an example?

`sim_checkpoint` in its current state here does raise the exceptions again after handling them, so I guess theoretically a user could do:

```python
try:
    sim_checkpoint(...)
except HealthCheckError as e:
    # ... additional handling ...

```

> Edit: actually maybe this adding of callback routines to the stepper has had a bit a scope creep, too. refactoring the checkpoint junk is a good deal of the business in #257.

I'm not sure if the original goal of this PR was to add a basic health check or just the callback functionality and nothing else, but if we want to add a health check here I think we have to deal with at least some of the stuff from #257 to make it work.

Reply:
> > Is there a good reason that `sim_checkpoint` does not make a crash viz dump at exception time for serial runs?
> 
> ?

Never mind, I see it now:

```python
    if comm is None:
        if terminate:
            raise exception
        return
```

Yeah, need a vis dump here @thomasgibson. Or maybe you could do something like:

```python
    if comm is not None:
        from mpi4py import MPI
        terminate = comm.allreduce(terminate, MPI.LOR)

    if terminate:
        # ...
```
to avoid multiple returns?

Reply:
> Sounds like we don't have a consensus about what we're actually trying to do here... i.e., are we trying to maximize user control, or minimize code? They conflict with each other to an extent. I was under the impression that we were aiming more for the latter.
> 

I agree with that.  My suggestion to add the callback hooks was centered on the need to give the user more freedom to implement what they want in the callbacks - one of those things likely to be health checking.  I wanted to let the user decide what to do once he had evaluated the state.

> I think it would make sense to have the user pass in a health check function to `sim_checkpoint`. I'm less sure I see the value in having them handle errors manually. Can you give an example?
> 

Health check is exactly the example, I think.  The user should be able to compose whatever health check they want.  In fact, lets say the user wants to monitor specific volume (because, for example, it tends to go wild at an injection inlet or something).  Then the user writes that health check function (or uses a built-in utility), and then ... if the check triggers, the user can either give up and quit, or turn on relaxation source terms to combat the issue.

> I'm not sure if the original goal of this PR was to add a basic health check or just the callback functionality and nothing else, but if we want to add a health check here I think we have to deal with at least some of the stuff from #257 to make it work.

It was my understanding that this PR was just trying to add the callbacks, however I see how its scope expanded, and I'm not necessarily opposed. I also want to make clear that my complaint about the structure and user control is not really a major one (from my perspective).  We can make use of it how it is, but it has evolved to the point that it doesn't really accomplish the things I had in mind for introducing the callbacks, and imo tries to do too much with automated health checking.